pkg/syncer/storage/storage_pv_controller.go

```diff
@@ -0,0 +1,184 @@
+*/
+
+package storage
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	kcpdynamic "github.com/kcp-dev/client-go/dynamic"
+	kcpdynamicinformer "github.com/kcp-dev/client-go/dynamic/dynamicinformer"
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/logging"
+	"github.com/kcp-dev/kcp/third_party/keyfunctions"
+)
+
+const persistentVolumeControllerName = "kcp-workload-syncer-storage-pv"
+
+type PersistentVolumeController struct {
+	queue                                 workqueue.RateLimitingInterface
+	upstreamInformers                     kcpdynamicinformer.DynamicSharedInformerFactory
+	upstreamClient                        kcpdynamic.ClusterInterface
+	downstreamClient                      dynamic.Interface
+	syncTarget                            syncTargetSpec
+	updateDownstreamPersistentVolumeClaim func(ctx context.Context, persistentVolumeClaim *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error)
+	getDownstreamPersistentVolumeClaim    func(persistentVolumeClaimName, persistentVolumeClaimNamespace string) (runtime.Object, error)
+	getUpstreamPersistentVolume           func(clusterName logicalcluster.Name, persistentVolumeName string) (runtime.Object, error)
+}
+
+// NewPersistentVolumeSyncer returns a new storage persistent volume syncer controller.
+func NewPersistentVolumeSyncer(
+	syncerLogger logr.Logger,
+	syncTargetWorkspace logicalcluster.Name,
+	syncTargetName, syncTargetKey string,
+	upstreamClient kcpdynamic.ClusterInterface,
+	downstreamClient dynamic.Interface,
+	downstreamKubeClient *kubernetes.Clientset,
+	upstreamInformers kcpdynamicinformer.DynamicSharedInformerFactory,
+	downstreamInformers dynamicinformer.DynamicSharedInformerFactory,
+	syncTargetUID types.UID,
+) (*PersistentVolumeController, error) {
+
+	c := &PersistentVolumeController{
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), persistentVolumeControllerName),
+		downstreamClient:  downstreamClient,
+		upstreamInformers: upstreamInformers,
+		upstreamClient:    upstreamClient,
+		syncTarget: syncTargetSpec{
+			name:      syncTargetName,
+			workspace: syncTargetWorkspace,
+			uid:       syncTargetUID,
+			key:       syncTargetKey,
+		},
+		updateDownstreamPersistentVolumeClaim: func(ctx context.Context, persistentVolumeClaim *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+			var err error
+			err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
I wouldn't do any retry here.
When there is an error, the main process function return the error and the resource is queued again.
cf https://github.com/kcp-dev/kcp/pull/2338/files#diff-59babfb03aea1b519e11dc21eaa7df1f4e810988bb0ccf4e96a6a0cd15bb26e5R177